lld/COFF/Writer.cpp

```diff
@@ -1413,6 +1415,36 @@ void Writer::removeUnusedSections() {
   llvm::erase_if(ctx.outputSections, isUnused);
 }
 
+void Writer::layoutSections() {
+  llvm::TimeTraceScope timeScope("Layout sections");
+  if (ctx.config.sectionOrder.empty())
+    return;
+
+  std::unordered_map<const OutputSection *, size_t> originalOrder;
+  for (size_t i = 0; i < ctx.outputSections.size(); ++i)
+    originalOrder[ctx.outputSections[i]] = i;
+
+  llvm::stable_sort(
+      ctx.outputSections,
+      [this, &originalOrder](const OutputSection *a, const OutputSection *b) {
+        auto itA = ctx.config.sectionOrder.find(a->name.str());
+        auto itB = ctx.config.sectionOrder.find(b->name.str());
+        bool aInOrder = itA != ctx.config.sectionOrder.end();
+        bool bInOrder = itB != ctx.config.sectionOrder.end();
+
+        // Put unspecified sections after all specified sections
+        if (aInOrder && bInOrder) {
+          return itA->second < itB->second;
+        } else if (aInOrder && !bInOrder) {
+          return true;
+        } else if (!aInOrder && bInOrder) {
+          return false;
+        } else {
+          return originalOrder[a] < originalOrder[b];
```

Comment:
Since this uses a stable sort, I think we also should be able to rely on that on keeping the original order of such elements, without needing to manually set up an `originalOrder` map. If we'd make this case (both elements unordered) into a plain `return false;`, we'd have `A < B -> false` and `B < A -> false`, meaning they are considered equivalent, and the stable sort should keep their original order - or am I misunderstanding something about stable sorts?